usr/sbin/p11sak/p11sak.c

```diff
@@ -1182,104 +1234,123 @@ static CK_RV print_vendor(CK_SESSION_HANDLE session, CK_OBJECT_HANDLE hkey, int
                 }
 
                 def_attr = 1;
-                
+
                 // print attribute by type
                 unsigned int hex = confignode_to_intval(hex_string)->value;
-                if (strcmp(confignode_to_bareval(type)->value, "CK_BBOOL") == 0) {
-                    // build template
-                    CK_BBOOL a_bool;
-                    CK_ATTRIBUTE temp = {hex, &a_bool, sizeof(a_bool)};
-
-                    // get attribute value
-                    rc = funcs->C_GetAttributeValue(session, hkey, &temp, 1);
-                    if (rc == CKR_OK) {
-                        if (temp.ulValueLen != sizeof(CK_BBOOL)) {
-                            fprintf(stderr, " Error in retrieving Attribute %s: %lu\n",
-                                    confignode_to_bareval(name)->value, temp.ulValueLen);
-                        } else {
-                            if (long_print) {
-                                printf("          %s: %s\n", confignode_to_bareval(name)->value,
-                                        CK_BBOOL2a(*(CK_BBOOL*) temp.pValue));
-                            } else if (strcmp(CK_BBOOL2a(*(CK_BBOOL*) temp.pValue), "CK_TRUE") == 0) {
-                                printf(" 1 ");
-                                return CKR_OK;
-                            }
-                        }    
-                    } else if (rc == CKR_ATTRIBUTE_SENSITIVE && long_print) {
-                        printf("          %s: SENSITIVE\n", confignode_to_bareval(name)->value);
-                    } else if (rc != CKR_ATTRIBUTE_TYPE_INVALID) {
-                        return rc;
-                    } 
-                } else if (strcmp((confignode_to_bareval(type)->value), "CK_ULONG") == 0) {
-                    // build template
-                    CK_ULONG a_ulong;
-                    CK_ATTRIBUTE temp = {hex, &a_ulong, sizeof(a_ulong)};
-
-                    // get attribute value
-                    rc = funcs->C_GetAttributeValue(session, hkey, &temp, 1);
-                    if (rc == CKR_OK) {
-                        if (temp.ulValueLen != sizeof(CK_ULONG)) {
-                            fprintf(stderr, " Error in retrieving Attribute %s: %lu\n",
-                                    confignode_to_bareval(name)->value, temp.ulValueLen);
-                        } else {
-                            if (long_print) {
-                                printf("          %s: %lu (0x%lX)\n", confignode_to_bareval(name)->value, 
-                                        *(CK_ULONG*) temp.pValue, *(CK_ULONG*) temp.pValue);    
+                if (!attr_already_printed(hex, attrs, attrs_len)) {
```

Comment:
So what do I have to do to get the same attribute printed twice?  I do not see any statement that I can only print the value once.  Especially since I can use different names for the same hex value I doubt this check is a good idea.

Reply:
Hmm ... different names for the same hex value. Why would a customer do this? With the old code I can reproduce duplicate printing of the same attribute (hex value) and I agree with Karel that this doesn't look good. However, we could leave it up to the customer to fix the config file instead of ignoring duplicate attrs in our code.

Reply:
Point is: this is a user configuration file.  We do not know why a user added the attribute twice.  Just because it "doesn't look good" does not mean that it is wrong.  If there is no real reason why a user should not be allowed to do it, why add that undocumented constraint?

Reply:
OK, here I would leave the decision up to Ingo: I can completely remove the commit that causes duplicate conf entries to be multiply written. Then we would just fix the config file handling.

Reply:
Maybe it would make sense to print the id the user defines in the config file too and let him use it multiple times. When it is on purpose he can do it and if it was an accident, because he didn't know it was already defined, he can tell by the printed id, even when he defined different names for the same id... to identify an id which was used multiple times maybe it would be good to sort the attributes in the long output by the id, so that the id's with multiple definitions are listed directly under each other. This would not be easy though with the current implementation, because an attribute is printed instantly after parsing, without knowing which attribute follows....

Reply:
Actually I also asked myself if rejecting a duplicate definition is really a good idea. I have not finally may up my mind though ....

Yes, it looks bad if you see duplicate definitions, but well, its a user config file, if the user does not like the result of what he configured, then he is free to change it :-) 

For simplicity, I guess I would tend to just not check for duplicate definitions.

Reply:
OK agreed: so we decide to just fix the config file handling. I removed the duplicate entries commit together with the whitespace fixes. We may do some more general housekeeping on p11sak later.